quickwit-metastore/src/metastore/file_backed_metastore/mod.rs

```diff
@@ -239,44 +269,109 @@ impl FileBackedMetastore {
 impl Metastore for FileBackedMetastore {
     /// -------------------------------------------------------------------------------
     /// Mutations over the high-level index.
-
     async fn create_index(&self, index_metadata: IndexMetadata) -> MetastoreResult<()> {
+        let index_id = index_metadata.index_id.clone();
+
         // We pick the outer lock here, so that we enter a critical section.
         let mut per_index_metastores_wlock = self.per_index_metastores.write().await;
 
-        let exists = per_index_metastores_wlock.contains_key(&index_metadata.index_id)
-            || index_exists(&*self.storage, &index_metadata.index_id).await?;
-
-        if exists {
-            return Err(MetastoreError::IndexAlreadyExists {
-                index_id: index_metadata.index_id.clone(),
+        // Checking if index already exists is a bit tedious:
+        // - first we check the index state: if it's `Alive`, return `IndexAlreadyExists` error, and
+        //   if it's `Creating` or `Deleting`, it's ok to override them as these are transitioning
+        //   states.
+        // - if the index is not in the indexes states map, we still need to check the storage as we
+        //   don't want to override an existing metadata file.
+        if let Some(index_state) = per_index_metastores_wlock.get(&index_id) {
+            if let IndexState::Alive(_) = index_state {
+                return Err(MetastoreError::IndexAlreadyExists {
+                    index_id: index_metadata.index_id.clone(),
+                });
+            }
+        } else if index_exists(&*self.storage, &index_metadata.index_id).await? {
+            return Err(MetastoreError::InternalError {
+                message: format!("Index {index_id} cannot be created."),
+                cause: anyhow::anyhow!(
+                    "Index {index_id} is not present in the `indexes_states.json` file but its \
+                     file `{index_id}/metastore.json` is on the storage."
+                ),
             });
         }
 
-        let index = FileBackedIndex::from(index_metadata);
+        // Set state to Creating` and rollback on metastore error.
+        per_index_metastores_wlock.insert(index_id.clone(), IndexState::Creating);
+        put_indexes_states(&*self.storage, &per_index_metastores_wlock)
+            .await
+            .map_err(|error| {
+                per_index_metastores_wlock.remove(&index_id);
```

Comment:
nitpick:
I think it is unexpected to put side-effects in a `map` function.

I prefer a if let Err for that.
```
if let Err(err) {
   er_index_metastores_wlock.remove(&index_id);
   return Err(err);
}
```

Reply:
Let's do the if let.